src/main/java/net/glowstone/net/handler/play/player/PlayerAbilitiesHandler.java

```diff
@@ -15,7 +18,20 @@ public void handle(GlowSession session, PlayerAbilitiesMessage message) {
 
         GlowPlayer player = session.getPlayer();
         boolean flying = (message.getFlags() & 0x02) != 0;
-
-        player.setFlying(player.getAllowFlight() && flying);
+        boolean isFlying = player.isFlying();
+        boolean canFly = player.getAllowFlight();
+        if (isFlying != flying) {
+            if (!flying || canFly) {
+                PlayerToggleFlightEvent event = EventFactory.getInstance()
+                        .callEvent(new PlayerToggleFlightEvent(player, flying));
+                boolean creative = player.getGameMode() == GameMode.CREATIVE;
+                int flags = (creative ? 8 : 0) | (canFly ? 4 : 0)
```

Comment:
Use the `sendAbilities` method inside `GlowPlayer` to avoid duplicating logic/code.

If the method is private, consider making an utility method for it. Could be in `GlowServer` for example (`sendPlayerAbilities(GlowPlayer)`)